base_layer/core/src/mempool/sync_protocol/mod.rs

```diff
@@ -0,0 +1,614 @@
+        task::spawn(async move {
+            let framed = framing::canonical(substream, MAX_FRAME_SIZE);
+            let protocol = MempoolPeerProtocol::new(framed, node_id.clone(), mempool);
+            match protocol.start_responder().await {
+                Ok(_) => {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Mempool responder protocol succeeded for peer `{}`",
+                        node_id.short_str()
+                    );
+                },
+                Err(err) => {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Mempool responder protocol failed for peer `{}`: {}",
+                        node_id.short_str(),
+                        err
+                    );
+                },
+            }
+        });
+    }
+}
+
+struct MempoolPeerProtocol<TSubstream, B> {
+    framed: Option<CanonicalFraming<TSubstream>>,
+    mempool: Mempool<B>,
+    peer_node_id: NodeId,
+}
+
+impl<TSubstream, B> MempoolPeerProtocol<TSubstream, B>
+where
+    TSubstream: AsyncRead + AsyncWrite + Unpin,
+    B: BlockchainBackend + 'static,
+{
+    pub fn new(framed: CanonicalFraming<TSubstream>, peer_node_id: NodeId, mempool: Mempool<B>) -> Self {
+        Self {
+            framed: Some(framed),
+            peer_node_id,
+            mempool,
+        }
+    }
+
+    pub async fn start_initiator(mut self) -> Result<(), MempoolProtocolError> {
+        let mut framed = self
+            .framed
+            .take()
+            .expect("MempoolProtocolError initialized without `framed`");
+        match self.start_initiator_inner(&mut framed).await {
+            Ok(_) => {
+                debug!(target: LOG_TARGET, "Initiator protocol complete");
+                Ok(())
+            },
+            Err(err) => {
+                if let Err(err) = framed.flush().await {
+                    debug!(target: LOG_TARGET, "IO error when flushing stream: {}", err);
+                }
+                if let Err(err) = framed.close().await {
+                    debug!(target: LOG_TARGET, "IO error when closing stream: {}", err);
+                }
+                Err(err)
+            },
+        }
+    }
+
+    async fn start_initiator_inner(
+        self,
+        framed: &mut CanonicalFraming<TSubstream>,
+    ) -> Result<(), MempoolProtocolError>
+    {
+        debug!(
+            target: LOG_TARGET,
+            "Starting initiator mempool sync for peer `{}`",
+            self.peer_node_id.short_str()
+        );
+
+        let transactions = async_mempool::snapshot(self.mempool.clone()).await?;
+        let key_len = <RistrettoSecretKey as SecretKey>::key_length();
+        let excess_sigs = transactions
+            .iter()
+            .map(|txn| txn.body.kernels()[0].excess_sig.get_signature())
+            // Never send a message > the frame size
+            .take(MAX_FRAME_SIZE / key_len)
```

Comment:
Ye I guess if we've got 98303 transactions in our mempool, we don't need to sync with a peer on startup? Remember this is sending the 32byte kernel excess sig of the transaction so that the receiving node doesn't send transactions that we already know about (i.e exclude txns listed in our inventory) - so if this message is full we _could_ start to get duplicate transactions. I guess a much more reasonable approach may be to just put a limit on how many transactions to sync.